worlds/heretic/__init__.py

```diff
@@ -0,0 +1,287 @@
+
+            loc = Locations.location_table[self.location_name_to_id[loc_name]]
+            if not self.included_episodes[loc["episode"] - 1]:
+                continue
+
+            self.multiworld.get_location(loc_name, self.player).place_locked_item(self.create_item(item_name))
+            self.location_count -= 1
+
+        # Give starting levels right away
+        for i in range(len(self.included_episodes)):
+            if self.included_episodes[i]:
+                self.multiworld.push_precollected(self.create_item(self.starting_level_for_episode[i]))
+        
+        # Give Computer area maps if option selected
+        if getattr(self.multiworld, "start_with_map_scrolls")[self.player].value:
+            for item_id, item_dict in Items.item_table.items():
+                item_episode = item_dict["episode"]
+                if item_episode > 0:
+                    if item_dict["doom_type"] == HERETIC_TYPE_MAP_SCROLL and self.included_episodes[item_episode - 1]:
+                        self.multiworld.push_precollected(self.create_item(item_dict["name"]))
+
+        # Fill the rest starting with powerups, then fillers
+        self.create_ratioed_items("Chaos Device", itempool)
+        self.create_ratioed_items("Morph Ovum", itempool)
+        self.create_ratioed_items("Mystic Urn", itempool)
+        self.create_ratioed_items("Ring of Invincibility", itempool)
+        self.create_ratioed_items("Shadowsphere", itempool)
+        self.create_ratioed_items("Timebomb of the Ancients", itempool)
+        self.create_ratioed_items("Tome of Power", itempool)
+        self.create_ratioed_items("Silver Shield", itempool)
+        self.create_ratioed_items("Enchanted Shield", itempool)
+
+        while len(itempool) < self.location_count:
+            itempool.append(self.create_item(self.get_filler_item_name()))
+
+        # add itempool to multiworld
+        self.multiworld.itempool += itempool
+
+    def get_filler_item_name(self):
+        return self.multiworld.random.choice([
+            "Quartz Flask",
+            "Crystal Geode",
+            "Energy Orb",
+            "Greater Runes",
+            "Inferno Orb",
+            "Pile of Mace Spheres",
+            "Quiver of Ethereal Arrows"
+        ])
+
+    def create_ratioed_items(self, item_name: str, itempool: List[HereticItem]):
+        remaining_loc = self.location_count - len(itempool)
+        if remaining_loc <= 0:
+            return
+
+        episode_count = self.get_episode_count()
+        count = min(remaining_loc, max(1, self.items_ratio[item_name] * episode_count))
+        if count == 0:
+            logger.warning("Warning, no " + item_name + " will be placed.")
+            return
+
+        for i in range(count):
+            itempool.append(self.create_item(item_name))
+
+    def fill_slot_data(self) -> Dict[str, Any]:
+        slot_data = {name: getattr(self.multiworld, name)[self.player].value for name in self.option_definitions}
```

Comment:
Do you really need all option results, including things like allow death logic?

Reply:
No, I'll do like I had to do for DOOM II:
```python
        return self.options.as_dict("difficulty", "random_monsters", "random_pickups", "random_music", "flip_levels", "allow_death_logic", "pro", "death_link", "reset_level_on_death", "episode1", "episode2", "episode3", "episode4")```